lib/bson/decimal128.rb

```diff
@@ -0,0 +1,307 @@
+    # are preserved but trailing zeroes may be lost.
+    # See the following example:
+    #
+    # @example
+    #  decimal128 = BSON::Decimal128.new("0.200")
+    #    => BSON::Decimal128('0.200')
+    #  big_decimal = decimal128.to_big_decimal
+    #    => #<BigDecimal:7fc619c95388,'0.2E0',9(18)>
+    #  big_decimal.to_s
+    #    => "0.2E0"
+    #
+    # Note that the the BSON::Decimal128 object can represent -NaN, sNaN,
+    # and -sNaN while Ruby's BigDecimal cannot.
+    #
+    # @return [ BigDecimal ] The decimal as a BigDecimal.
+    #
+    # @since 4.2.0
+    def to_big_decimal
+      @big_decimal ||= BigDecimal.new(to_s)
+    end
+
+    private
+
+    def set_bits(low, high)
+      @low = low
+      @high = high
+    end
+
+    class << self
+
+      # Deserialize the decimal128 from raw BSON bytes.
+      #
+      # @example Get the decimal128 from BSON.
+      #   Decimal128.from_bson(bson)
+      #
+      # @param [ ByteBuffer ] buffer The byte buffer.
+      #
+      # @return [ BSON::Decimal128 ] The decimal object.
+      #
+      # @since 4.2.0
+      def from_bson(buffer)
+        from_bits(*buffer.get_decimal128_bytes.unpack('Q*'))
+      end
+
+      # Instantiate a Decimal128 from a string.
+      #
+      # @example Create a Decimal128 from a string.
+      #   BSON::Decimal128.from_string("1.05E+3")
+      #
+      # @param [ String ] string The string to parse.
+      #
+      # @raise [ BSON::Decimal128::InvalidString ] If the provided string is invalid.
+      #
+      # @return [ BSON::Decimal128 ] The new decimal128.
+      #
+      # @since 4.2.0
+      def from_string(string)
+        from_bits(*Builder::FromString.new(string).bits)
+      end
+
+      # Instantiate a Decimal128 from a BigDecimal.
+      #
+      # @example Create a Decimal128 object from a BigDecimal.
+      #   BSON::Decimal128.from_big_decimal(big_decimal)
+      #
+      # @param [ BigDecimal ] big_decimal The big decimal to instantiate the
+      #   Decimal128 from.
+      #
+      # @raise [ BSON::Decimal128::InvalidBigDecimal ] If the provided value is not a
+      #   BigDecimal.
+      #
+      # @return [ BSON::Decimal128 ] The new decimal128.
+      #
+      # @since 4.2.0
+      def from_big_decimal(big_decimal)
+        new(big_decimal)
+      end
+
+      # Instantiate a Decimal128 from a high and low bits.
+      #
+      # @example Create a Decimal128 from high and low bits.
+      #   BSON::Decimal128.from_bits(high, low)
+      #
+      # @param [ Integer ] high The high order bits.
+      # @param [ Integer ] low The low order bits.
+      #
+      # @return [ BSON::Decimal128 ] The new decimal128.
+      #
+      # @since 4.2.0
+      def from_bits(low, high)
+        decimal = allocate
+        decimal.send(:set_bits, low, high)
+        decimal
+      end
+    end
+
+    # Raised when a Decimal128 is instantiated with a non-BigDecimal type.
+    #
+    # @since 4.2.0
+    class InvalidBigDecimal < RuntimeError; end
```

Comment:
Perhaps...but I can't change InvalidString or InvalidRange to extend ArgumentError until a major version of the bson gem. I'm getting rid of this error anyway and letting BigDecimal raise if it cannot be instantiated with the args to #initialize.